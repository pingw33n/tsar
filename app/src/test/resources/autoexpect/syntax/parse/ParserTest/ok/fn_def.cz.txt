@(0..563) {
  source: "parseOk/fn_def.cz"
  items: [
    FnDef @(0..11) {
      name: @(3..6) `foo`
    }
    FnDef @(12..27) {
      pub: @(12..15) {
      }
      name: @(19..22) `foo`
    }
    FnDef @(28..46) {
      name: @(38..41) `foo`
      unsafe: @(28..34)
    }
    FnDef @(47..69) {
      pub: @(47..50) {
      }
      name: @(61..64) `foo`
      unsafe: @(51..57)
    }
    FnDef @(70..104) {
      pub: @(70..73) {
      }
      name: @(84..87) `foo`
      params: [
        @(88..93) {
          name: @(88..89) `s`
          type: Path @(91..93) {
            @(91..93) {
              suffix: [
                Item @(91..93) {
                  item: @(91..93) {
                    name: @(91..93) `u8`
                  }
                }
              ]
            }
          }
        }
      ]
      result: Path @(98..101) {
        @(98..101) {
          suffix: [
            Item @(98..101) {
              item: @(98..101) {
                name: @(98..101) `i32`
              }
            }
          ]
        }
      }
      unsafe: @(74..80)
    }
    FnDef @(106..133) {
      pub: @(106..109) {
      }
      name: @(117..121) `main`
    }
    FnDef @(135..350) {
      name: @(138..145) `foo_bar`
      typeParams: [
        @(151..152) `T`
        @(158..159) `U`
        @(165..166) `V`
      ]
      params: [
        @(184..200) {
          name: @(184..186) `p1`
          type: Path @(197..200) {
            @(197..200) {
              suffix: [
                Item @(197..200) {
                  item: @(197..200) {
                    name: @(197..200) `u32`
                  }
                }
              ]
            }
          }
        }
        @(211..282) {
          label: @(211..217) `label2`
          name: @(222..224) `p2`
          type: Path @(235..282) {
            @(235..282) {
              prefix: [
                @(235..238) {
                  name: @(235..238) `std`
                }
              ]
              suffix: [
                Item @(250..282) {
                  item: @(250..282) {
                    name: @(250..251) `T`
                    typeArgs: [
                      Path @(262..282) {
                        @(262..282) {
                          suffix: [
                            Item @(262..282) {
                              item: @(262..282) {
                                name: @(262..263) `U`
                                typeArgs: [
                                  Path @(274..275) {
                                    @(274..275) {
                                      suffix: [
                                        Item @(274..275) {
                                          item: @(274..275) {
                                            name: @(274..275) `V`
                                          }
                                        }
                                      ]
                                    }
                                  }
                                ]
                              }
                            }
                          ]
                        }
                      }
                    ]
                  }
                }
              ]
            }
          }
        }
        @(288..319) {
          label: @(288..289) `_`
          name: @(290..292) `p3`
          type: Path @(294..319) {
            @(294..319) {
              origin: Super @(294..308) {
                count: `2`
              }
              prefix: [
                @(308..314) {
                  name: @(308..311) `bar`
                  typeArgs: [
                    Path @(312..313) {
                      @(312..313) {
                        suffix: [
                          Item @(312..313) {
                            item: @(312..313) {
                              name: @(312..313) `T`
                            }
                          }
                        ]
                      }
                    }
                  ]
                }
              ]
              suffix: [
                Item @(316..319) {
                  item: @(316..319) {
                    name: @(316..317) `T`
                  }
                }
              ]
            }
          }
        }
      ]
      result: Path @(330..346) {
        @(330..346) {
          origin: Package @(330..339) {
          }
          prefix: [
            @(339..343) {
              name: @(339..343) `test`
            }
          ]
          suffix: [
            Item @(345..346) {
              item: @(345..346) {
                name: @(345..346) `R`
              }
            }
          ]
        }
      }
    }
    FnDef @(352..374) {
      name: @(355..358) `foo`
      params: [
        @(359..364) {
          name: @(359..360) `a`
          type: UnnamedStruct @(362..364) {
          }
        }
      ]
      result: UnnamedStruct @(369..371) {
      }
    }
    FnDef @(375..403) {
      name: @(378..381) `foo`
      params: [
        @(382..391) {
          name: @(382..383) `a`
          type: UnnamedStruct @(385..391) {
            fields: [
              {
                name: @(386..389) `0`
                type: Path @(386..389) {
                  @(386..389) {
                    suffix: [
                      Item @(386..389) {
                        item: @(386..389) {
                          name: @(386..389) `u32`
                        }
                      }
                    ]
                  }
                }
              }
            ]
          }
        }
      ]
      result: UnnamedStruct @(396..400) {
        fields: [
          {
            name: @(397..398) `0`
            type: Path @(397..398) {
              @(397..398) {
                suffix: [
                  Item @(397..398) {
                    item: @(397..398) {
                      name: @(397..398) `T`
                    }
                  }
                ]
              }
            }
          }
        ]
      }
    }
    FnDef @(404..442) {
      name: @(407..410) `foo`
      params: [
        @(411..422) {
          name: @(411..412) `a`
          type: UnnamedStruct @(414..422) {
            fields: [
              {
                name: @(415..416) `A`
                type: Path @(418..421) {
                  @(418..421) {
                    suffix: [
                      Item @(418..421) {
                        item: @(418..421) {
                          name: @(418..421) `u32`
                        }
                      }
                    ]
                  }
                }
              }
            ]
          }
        }
      ]
      result: UnnamedStruct @(427..439) {
        fields: [
          {
            name: @(428..429) `b`
            type: Path @(431..432) {
              @(431..432) {
                suffix: [
                  Item @(431..432) {
                    item: @(431..432) {
                      name: @(431..432) `T`
                    }
                  }
                ]
              }
            }
          }
          {
            name: @(434..435) `c`
            type: Path @(437..438) {
              @(437..438) {
                suffix: [
                  Item @(437..438) {
                    item: @(437..438) {
                      name: @(437..438) `D`
                    }
                  }
                ]
              }
            }
          }
        ]
      }
    }
    FnDef @(443..461) {
      name: @(446..449) `foo`
      params: [
        @(450..457) {
          name: @(450..451) `s`
          type: Ref @(453..457) {
            Path @(454..457) {
              @(454..457) {
                suffix: [
                  Item @(454..457) {
                    item: @(454..457) {
                      name: @(454..457) `str`
                    }
                  }
                ]
              }
            }
          }
        }
      ]
    }
    FnDef @(463..547) {
      name: @(466..467) `f`
      params: [
        @(471..475) {
          label: @(471..475) `self`
          name: @(471..475) `self`
          type: Path @(471..475) {
            @(471..475) {
              prefix: [
                @(471..475) {
                  name: @(471..475) `Self`
                }
              ]
            }
          }
        }
        @(479..493) {
          label: @(479..480) `_`
          name: @(481..488) `unnamed`
          type: Path @(490..493) {
            @(490..493) {
              suffix: [
                Item @(490..493) {
                  item: @(490..493) {
                    name: @(490..493) `u32`
                  }
                }
              ]
            }
          }
        }
        @(497..506) {
          name: @(497..501) `arg1`
          type: Path @(503..506) {
            @(503..506) {
              suffix: [
                Item @(503..506) {
                  item: @(503..506) {
                    name: @(503..506) `u32`
                  }
                }
              ]
            }
          }
        }
        @(510..522) {
          label: @(510..512) `fn`
          name: @(513..517) `name`
          type: Path @(519..522) {
            @(519..522) {
              suffix: [
                Item @(519..522) {
                  item: @(519..522) {
                    name: @(519..522) `u32`
                  }
                }
              ]
            }
          }
        }
        @(526..541) {
          label: @(526..527) `_`
          name: @(528..536) `unnamed2`
          type: Path @(538..541) {
            @(538..541) {
              suffix: [
                Item @(538..541) {
                  item: @(538..541) {
                    name: @(538..541) `u32`
                  }
                }
              ]
            }
          }
        }
      ]
    }
    FnDef @(549..563) {
      name: @(552..553) `f`
      params: [
        @(554..559) {
          label: @(555..559) `self`
          name: @(555..559) `self`
          type: Ref @(554..559) {
            Path @(555..559) {
              @(555..559) {
                prefix: [
                  @(555..559) {
                    name: @(555..559) `Self`
                  }
                ]
              }
            }
          }
        }
      ]
    }
  ]
}
