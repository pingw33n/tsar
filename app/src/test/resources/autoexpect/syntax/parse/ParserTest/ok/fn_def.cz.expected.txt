@(0..673) {
  source: "parseOk/fn_def.cz"
  items: [
    FnDef @(0..11) {
      name: @(3..6) `foo`
    }
    FnDef @(12..27) {
      pub: @(12..15) {
      }
      name: @(19..22) `foo`
    }
    FnDef @(28..46) {
      name: @(38..41) `foo`
      unsafe: @(28..34)
    }
    FnDef @(47..69) {
      pub: @(47..50) {
      }
      name: @(61..64) `foo`
      unsafe: @(51..57)
    }
    FnDef @(70..104) {
      pub: @(70..73) {
      }
      name: @(84..87) `foo`
      params: [
        @(88..93) {
          label: @(88..89) `s`
          name: @(88..89) `s`
          type: Path @(91..93) {
            @(91..93) {
              suffix: [
                Item @(91..93) {
                  item: @(91..93) {
                    name: @(91..93) `u8`
                  }
                }
              ]
            }
          }
        }
      ]
      result: Path @(98..101) {
        @(98..101) {
          suffix: [
            Item @(98..101) {
              item: @(98..101) {
                name: @(98..101) `i32`
              }
            }
          ]
        }
      }
      unsafe: @(74..80)
    }
    FnDef @(106..133) {
      pub: @(106..109) {
      }
      name: @(117..121) `main`
    }
    FnDef @(135..350) {
      name: @(138..145) `foo_bar`
      typeParams: [
        @(151..152) `T`
        @(158..159) `U`
        @(165..166) `V`
      ]
      params: [
        @(184..200) {
          label: @(184..186) `p1`
          name: @(184..186) `p1`
          type: Path @(197..200) {
            @(197..200) {
              suffix: [
                Item @(197..200) {
                  item: @(197..200) {
                    name: @(197..200) `u32`
                  }
                }
              ]
            }
          }
        }
        @(211..282) {
          label: @(211..217) `label2`
          name: @(222..224) `p2`
          type: Path @(235..282) {
            @(235..282) {
              prefix: [
                @(235..238) {
                  name: @(235..238) `std`
                }
              ]
              suffix: [
                Item @(250..282) {
                  item: @(250..282) {
                    name: @(250..251) `T`
                    typeArgs: [
                      Path @(262..282) {
                        @(262..282) {
                          suffix: [
                            Item @(262..282) {
                              item: @(262..282) {
                                name: @(262..263) `U`
                                typeArgs: [
                                  Path @(274..275) {
                                    @(274..275) {
                                      suffix: [
                                        Item @(274..275) {
                                          item: @(274..275) {
                                            name: @(274..275) `V`
                                          }
                                        }
                                      ]
                                    }
                                  }
                                ]
                              }
                            }
                          ]
                        }
                      }
                    ]
                  }
                }
              ]
            }
          }
        }
        @(288..319) {
          label: @(288..289)
          name: @(290..292) `p3`
          type: Path @(294..319) {
            @(294..319) {
              origin: Super @(294..308) {
                count: `2`
              }
              prefix: [
                @(308..314) {
                  name: @(308..311) `bar`
                  typeArgs: [
                    Path @(312..313) {
                      @(312..313) {
                        suffix: [
                          Item @(312..313) {
                            item: @(312..313) {
                              name: @(312..313) `T`
                            }
                          }
                        ]
                      }
                    }
                  ]
                }
              ]
              suffix: [
                Item @(316..319) {
                  item: @(316..319) {
                    name: @(316..317) `T`
                  }
                }
              ]
            }
          }
        }
      ]
      result: Path @(330..346) {
        @(330..346) {
          origin: Package @(330..339) {
          }
          prefix: [
            @(339..343) {
              name: @(339..343) `test`
            }
          ]
          suffix: [
            Item @(345..346) {
              item: @(345..346) {
                name: @(345..346) `R`
              }
            }
          ]
        }
      }
    }
    FnDef @(352..374) {
      name: @(355..358) `foo`
      params: [
        @(359..364) {
          label: @(359..360) `a`
          name: @(359..360) `a`
          type: UnnamedStruct @(362..364) {
          }
        }
      ]
      result: UnnamedStruct @(369..371) {
      }
    }
    FnDef @(375..403) {
      name: @(378..381) `foo`
      params: [
        @(382..391) {
          label: @(382..383) `a`
          name: @(382..383) `a`
          type: UnnamedStruct @(385..391) {
            fields: [
              {
                name: @(386..389) `0`
                type: Path @(386..389) {
                  @(386..389) {
                    suffix: [
                      Item @(386..389) {
                        item: @(386..389) {
                          name: @(386..389) `u32`
                        }
                      }
                    ]
                  }
                }
              }
            ]
          }
        }
      ]
      result: UnnamedStruct @(396..400) {
        fields: [
          {
            name: @(397..398) `0`
            type: Path @(397..398) {
              @(397..398) {
                suffix: [
                  Item @(397..398) {
                    item: @(397..398) {
                      name: @(397..398) `T`
                    }
                  }
                ]
              }
            }
          }
        ]
      }
    }
    FnDef @(404..442) {
      name: @(407..410) `foo`
      params: [
        @(411..422) {
          label: @(411..412) `a`
          name: @(411..412) `a`
          type: UnnamedStruct @(414..422) {
            fields: [
              {
                name: @(415..416) `A`
                type: Path @(418..421) {
                  @(418..421) {
                    suffix: [
                      Item @(418..421) {
                        item: @(418..421) {
                          name: @(418..421) `u32`
                        }
                      }
                    ]
                  }
                }
              }
            ]
          }
        }
      ]
      result: UnnamedStruct @(427..439) {
        fields: [
          {
            name: @(428..429) `b`
            type: Path @(431..432) {
              @(431..432) {
                suffix: [
                  Item @(431..432) {
                    item: @(431..432) {
                      name: @(431..432) `T`
                    }
                  }
                ]
              }
            }
          }
          {
            name: @(434..435) `c`
            type: Path @(437..438) {
              @(437..438) {
                suffix: [
                  Item @(437..438) {
                    item: @(437..438) {
                      name: @(437..438) `D`
                    }
                  }
                ]
              }
            }
          }
        ]
      }
    }
    FnDef @(443..461) {
      name: @(446..449) `foo`
      params: [
        @(450..457) {
          label: @(450..451) `s`
          name: @(450..451) `s`
          type: Ref @(453..457) {
            Path @(454..457) {
              @(454..457) {
                suffix: [
                  Item @(454..457) {
                    item: @(454..457) {
                      name: @(454..457) `str`
                    }
                  }
                ]
              }
            }
          }
        }
      ]
    }
    FnDef @(463..546) {
      name: @(466..467) `f`
      params: [
        @(471..475) {
          label: @(471..475) `self`
          name: @(471..475) `self`
          type: Path @(471..475) {
            @(471..475) {
              suffix: [
                Item @(471..475) {
                  item: @(471..475) {
                    name: @(471..475) `Self`
                  }
                }
              ]
            }
          }
        }
        @(479..493) {
          label: @(479..480)
          name: @(481..488) `unnamed`
          type: Path @(490..493) {
            @(490..493) {
              suffix: [
                Item @(490..493) {
                  item: @(490..493) {
                    name: @(490..493) `u32`
                  }
                }
              ]
            }
          }
        }
        @(497..506) {
          label: @(497..501) `arg1`
          name: @(497..501) `arg1`
          type: Path @(503..506) {
            @(503..506) {
              suffix: [
                Item @(503..506) {
                  item: @(503..506) {
                    name: @(503..506) `u32`
                  }
                }
              ]
            }
          }
        }
        @(510..521) {
          label: @(510..511) `f`
          name: @(512..516) `name`
          type: Path @(518..521) {
            @(518..521) {
              suffix: [
                Item @(518..521) {
                  item: @(518..521) {
                    name: @(518..521) `u32`
                  }
                }
              ]
            }
          }
        }
        @(525..540) {
          label: @(525..526)
          name: @(527..535) `unnamed2`
          type: Path @(537..540) {
            @(537..540) {
              suffix: [
                Item @(537..540) {
                  item: @(537..540) {
                    name: @(537..540) `u32`
                  }
                }
              ]
            }
          }
        }
      ]
    }
    FnDef @(548..562) {
      name: @(551..552) `f`
      params: [
        @(553..558) {
          label: @(554..558) `self`
          name: @(554..558) `self`
          type: Ref @(553..558) {
            Path @(554..558) {
              @(554..558) {
                suffix: [
                  Item @(554..558) {
                    item: @(554..558) {
                      name: @(554..558) `Self`
                    }
                  }
                ]
              }
            }
          }
        }
      ]
    }
    FnDef @(564..673) {
      name: @(567..568) `f`
      params: [
        @(569..582) {
          label: @(569..572) `foo`
          name: @(569..572) `foo`
          type: Path @(574..577) {
            @(574..577) {
              suffix: [
                Item @(574..577) {
                  item: @(574..577) {
                    name: @(574..577) `u32`
                  }
                }
              ]
            }
          }
          default: FnCall @(580..582) {
          }
        }
        @(584..619) {
          label: @(584..587) `bar`
          name: @(584..587) `bar`
          type: Path @(589..592) {
            @(589..592) {
              suffix: [
                Item @(589..592) {
                  item: @(589..592) {
                    name: @(589..592) `Foo`
                  }
                }
              ]
            }
          }
          default: BinaryOp @(595..619) {
            kind: @(602..603) `ADD`
            left: Int @(595..597) {
              `42`
            }
            right: BinaryOp @(604..619) {
              kind: @(612..613) `DIV`
              left: Int @(604..607) {
                `100`
              }
              right: Int @(614..619) {
                `10000`
              }
            }
          }
        }
        @(630..653) {
          label: @(630..631)
          name: @(632..635) `baz`
          type: Path @(637..638) {
            @(637..638) {
              suffix: [
                Item @(637..638) {
                  item: @(637..638) {
                    name: @(637..638) `T`
                  }
                }
              ]
            }
          }
          default: Block @(644..653) {
            [
              Expr {
                Path @(646..647) {
                  @(646..647) {
                    suffix: [
                      Item @(646..647) {
                        item: @(646..647) {
                          name: @(646..647) `a`
                        }
                      }
                    ]
                  }
                }
              }
              Expr {
                Path @(648..649) {
                  @(648..649) {
                    suffix: [
                      Item @(648..649) {
                        item: @(648..649) {
                          name: @(648..649) `b`
                        }
                      }
                    ]
                  }
                }
              }
              Expr {
                Path @(650..651) {
                  @(650..651) {
                    suffix: [
                      Item @(650..651) {
                        item: @(650..651) {
                          name: @(650..651) `c`
                        }
                      }
                    ]
                  }
                }
              }
            ]
          }
        }
        @(659..669) {
          label: @(659..660) `p`
          name: @(659..660) `p`
          type: Path @(662..663) {
            @(662..663) {
              suffix: [
                Item @(662..663) {
                  item: @(662..663) {
                    name: @(662..663) `T`
                  }
                }
              ]
            }
          }
          default: FnCall @(666..669) {
            callee: Free {
              Path @(666..667) {
                @(666..667) {
                  suffix: [
                    Item @(666..667) {
                      item: @(666..667) {
                        name: @(666..667) `f`
                      }
                    }
                  ]
                }
              }
            }
          }
        }
      ]
    }
  ]
}
